utils.py

```diff
@@ -289,3 +290,29 @@ def natural_sort(l):
     convert = lambda text: int(text) if text.isdigit() else text.lower() 
     alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ] 
     return sorted(l, key = alphanum_key)
+
+class FunctionCallOperation(Operation):
+    def __init__(self, x, callback):
+        super(FunctionCallOperation, self).__init__([x], x.mesh, name="FunctionCall")
+        self._outputs = [Tensor(self, x.shape, x.dtype)]
+        def wrapper(x):
+            callback(x)
+            return x
+        self._callback = wrapper
+
+    def lower(self, lowering):
+        tensors = lowering.tensors[self.inputs[0]]
+        tensors = tensors.to_laid_out_tensor().tensor_list
+
+        # hack to force the function onto the graph
+        tensors[0] = tf2.tuple([tensors[0]], control_inputs=[
+            tf.py_func(self._callback, [tensors], [tensors[0].dtype])[0]])[0]
+        
+        tensors = lowering.mesh_impl(self).LaidOutTensor(tensors)
+        lowering.set_tensor_lowering(self.outputs[0], tensors)
+
+    def gradient(self, g):
+        return g
+
+def hook_grah(x, callback):
```

Comment:
should we call this `hook_graph` instead of hook grah or is this spelling intentional?

Reply:
Definitely a spelling mistake, nice catch